Sources/Apodini/Semantic Model Builder/EndpointsTree.swift

```diff
@@ -0,0 +1,154 @@
+    let handleMethod: () -> ResponseEncodable
+    let responseTransformers: [() -> (AnyResponseTransformer)]
+    
+    /// Type returned by `handle()`
+    let handleReturnType: ResponseEncodable.Type
+    
+    /// Response type ultimately returned by `handle()` and possible following `ResponseTransformer`s
+    lazy var responseType: ResponseEncodable.Type = {
+        guard let lastResponseTransformer = self.responseTransformers.last else {
+            return self.handleReturnType
+        }
+        return lastResponseTransformer().transformedResponseType
+    }()
+    
+    /// All `@Parameter` `RequestInjectable`s that are used inside handling `Component`
+    lazy var parameters: [EndpointParameter] = EndpointParameter.create(from: requestInjectables)
+
+    lazy var pathComponents: [_PathComponent] = {
+        treeNode.pathComponents
+    }()
+}
+
+class EndpointsTreeNode {
+    let path: _PathComponent
+    lazy var pathComponents: [_PathComponent] = {
+        var paths: [_PathComponent] = []
+        collectPathComponents(pathComponents: &paths)
+        return paths
+    }()
+
+    var endpoints: [Operation: Endpoint] = [:]
+
+    let parent: EndpointsTreeNode?
+    private var nodeChildren: [String: EndpointsTreeNode] = [:]
+    var children: Dictionary<String, EndpointsTreeNode>.Values {
+        nodeChildren.values
+    }
+    /// If a EndpointsTreeNode A is a child to  a EndpointsTreeNode B and A has an `PathParameter` as its `path`
+    /// B can't have any other children besides A that also have an `PathParameter` at the same location.
+    /// Thus we mark `childContainsPathParameter` to true as soon as we insert a `PathParameter` as a child.
+    private var childContainsPathParameter = false
+
+    init(path: _PathComponent, parent: EndpointsTreeNode? = nil) {
+        self.path = path
+        self.parent = parent
+    }
+
+    func addEndpoint(_ endpoint: inout Endpoint, at paths: [PathComponent]) {
+        if paths.isEmpty {
+            // swiftlint:disable:next force_unwrapping
+            precondition(endpoints[endpoint.operation] == nil, "Tried overwriting endpoint \(endpoints[endpoint.operation]!.description) with \(endpoint.description) for operation \(endpoint.operation)")
+            precondition(endpoint.treeNode == nil, "The endpoint \(endpoint.description) is already inserted at some different place")
+
+            endpoint.treeNode = self
+            endpoints[endpoint.operation] = endpoint
+        } else {
+            var pathComponents = paths
+            if let first = pathComponents.removeFirst() as? _PathComponent {
+                var child = nodeChildren[first.description]
+                if child == nil {
+                    // swiftlint:disable:next todo
+                    // TODO is there a better way to check for type Parameter instead of string comparison?
+                    if let info = try? typeInfo(of: type(of: first)), info.mangledName == "Parameter" {
```

Comment:
The same would apply to this line here: https://github.com/Apodini/Apodini/blob/2c7649ed5cc1c59001c71f396c7b1f1862d180eb/Sources/Apodini/Semantic%20Model%20Builder/EndpointParameter.swift#L38

Reply:
Do I understand you correct, the idea is to add a line like you suggested which directly references `Parameter`, but **isn't used** actually, but serves as a hint when Parameter would change? Or what is the idea here?
Not sure if this solves our problem in the days of automatic refactoring? Most editors (like Xcode) probably support renaming all **used** references, only few IDEs search for string usage as well.

Reply:
As far as I've understood, we should do this to avoid the String Comparison:
```
let parameterType = try? typeInfo(of: Parameter<String>.self)
if let info = try? typeInfo(of: type(of: first)), info.mangledName == parameterType?.mangledName { 
/// ...do something
}
```

where String can be replaced by Int, doesn't matter.
I've tested it just now, it works and tests are still succeeding.

Reply:
ah yeah, my head went to a complete different direction 😂 
@lschlesinger do you want to make those changes to resolve this ☺️?

Reply:
solved with 6f809064ec38b67b5d3f9e435cf14bc115791f15